apps/ui/src/Button.tsx

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { cva, type VariantProps } from 'class-variance-authority';
+import { ButtonHTMLAttributes, forwardRef, type ReactNode } from 'react';
+
+import { LoadingIndicator } from './LoadingIndicator';
+import { Slot } from '@radix-ui/react-slot';
+
+const buttonStyles = cva(['inline-flex flex-nowrap items-center justify-center gap-2 relative'], {
+	variants: {
+		variant: {
+			primary: 'bg-sui-dark text-sui-light hover:text-white border-none',
+			secondary: 'bg-gray-90 text-gray-50 hover:text-white border-none',
+			outline:
+				'bg-white border border-steel text-steel-dark hover:text-steel-darker hover:border-steel-dark active:text-steel active:border-steel disabled:border-gray-45 disabled:text-steel-dark',
+		},
+		size: {
+			md: 'px-3 py-2 rounded-md text-bodySmall font-semibold',
+			lg: 'px-4 py-3 rounded-lg text-body font-semibold',
+		},
+		loading: {
+			true: 'text-transparent',
+		},
+	},
+	defaultVariants: {
+		variant: 'primary',
+		size: 'md',
+	},
+});
+
+export interface ButtonProps
+	extends VariantProps<typeof buttonStyles>,
+		ButtonHTMLAttributes<HTMLButtonElement> {
+	before?: ReactNode;
+	after?: ReactNode;
+	asChild?: boolean;
+}
+
+export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
+	({ variant, size, loading, children, before, after, asChild, ...props }, ref) => {
+		const Comp = asChild ? Slot : 'button';
+
+		if (asChild && (loading || before || after)) {
```

Comment:
Is there any benefit to enforcing this at the type level or too overkill maybe? This works fine though haha

Reply:
It's certainly possible, I was just thinking it's probably overkill, and I'm slowly getting convinced that consumers should just do the before / after themselves (loading is still iffy, but there's probably a clever way to do it without needing to compose the children layout).